ee/clickhouse/queries/trends/total_volume.py

```diff
@@ -38,48 +33,156 @@ def _total_volume_query(self, entity: Entity, filter: Filter, team_id: int) -> T
 
-        if filter.display in TRENDS_DISPLAY_BY_VALUE:
-            content_sql = VOLUME_TOTAL_AGGREGATE_SQL.format(event_query=event_query, **content_sql_params)
-            time_range = enumerate_time_range(filter, seconds_in_interval)
-
-            return (
-                content_sql,
-                params,
-                lambda result: [
-                    {"aggregated_value": result[0][0] if result and len(result) else 0, "days": time_range}
-                ],
+        if entity.math in [WEEKLY_ACTIVE, MONTHLY_ACTIVE]:
+            final_query = ACTIVE_USER_SQL.format(
+                event_query=event_query,
+                **content_sql_params,
+                parsed_date_to=trend_event_query.parsed_date_to,
+                parsed_date_from=trend_event_query.parsed_date_from,
+                **trend_event_query.active_user_params
             )
         else:
+            final_query = VOLUME_SQL.format(event_query=event_query, **content_sql_params)
 
-            if entity.math in [WEEKLY_ACTIVE, MONTHLY_ACTIVE]:
-                content_sql = ACTIVE_USER_SQL.format(
-                    event_query=event_query,
-                    **content_sql_params,
-                    parsed_date_to=trend_event_query.parsed_date_to,
-                    parsed_date_from=trend_event_query.parsed_date_from,
-                    **trend_event_query.active_user_params
-                )
-            else:
-                content_sql = VOLUME_SQL.format(event_query=event_query, **content_sql_params)
-
-            null_sql = NULL_SQL.format(trunc_func=trunc_func, interval_func=interval_func)
-            params["date_from"] = format_ch_timestamp(filter.date_from or get_earliest_timestamp(team_id), filter)
-            params["date_to"] = format_ch_timestamp(filter.date_to, filter)
-            params["interval"] = filter.interval
-            final_query = AGGREGATE_SQL.format(null_sql=null_sql, content_sql=content_sql)
-            return final_query, params, self._parse_total_volume_result(filter)
+        params["date_from"] = format_ch_timestamp(filter.date_from or get_earliest_timestamp(team_id), filter)
+        params["date_to"] = format_ch_timestamp(filter.date_to, filter)
+        params["interval"] = filter.interval
+        return final_query, params, self._parse_total_volume_result(filter)
 
     def _parse_total_volume_result(self, filter: Filter) -> Callable:
         def _parse(result: List) -> List:
-            parsed_results = []
-            for _, stats in enumerate(result):
-                parsed_result = parse_response(stats, filter)
-                parsed_results.append(parsed_result)
+            # NOTE: there's some implicit coupling here:
+            #
+            #        1. the ordering of the fields we get back, (date, value)
+            #        2. that the last row is always the `WITH TOTALS` result
+            #
+            #  TODO: remove coupling
```

Comment:
Who is this TODO for and is this actionable or relevant?

While there is coupling involved here of, so does every bit of querying and query postprocessing. Decoupling them makes things _worse_ more often than not unless the SQL is generic somehow.

Reply:
I’ll remove the todo, agree not particularly useful.

On the point of coupling, I think that’s fine as long as the couple is explicit and obvious. Even better if coupling can be picked up by type checker e.g. having an explicit interface defined like InsightResult. Still working on what that could look like though